torch/nn/modules/transformer.py

```diff
@@ -0,0 +1,290 @@
+import torch
+import copy
+from .. import functional as F
+from .module import Module
+from .activation import MultiheadAttention
+from .container import ModuleList
+import numpy as np
+from ..init import xavier_uniform_
+from .dropout import Dropout
+from .linear import Linear
+from .normalization import LayerNorm
+
+class TransformerBase(Module):
+    r"""A base Transformer class. The transformer is based on a standard
+        Encoder-Decoder architecture.
+
+    Args:
+        encoder: an encoder component (required).
+        decoder: a decoder component (required).
+
+    Examples:
+        >>> transformer_model = nn.TransformerBase(encoder, decoder)
+    """
+
+    def __init__(self, encoder, decoder):
+        super(TransformerBase, self).__init__()
+        self.encoder = encoder 
+        self.decoder = decoder
+
+
+class Transformer(TransformerBase):
+    r"""A transformer model. User is able to modified the attributes as needed.
+
+    Args:
+        d_model: the number of expected features in the encoder/decoder inputs (default=512).
+        nhead: the number of heads in the multiheadattention models (default=8).
+        num_encoder_layers: the number of sub-encoder-layers in the encoder (default=6).
+        num_decoder_layers: the number of sub-decoder-layers in the decoder (default=6).
+        dim_feedforward: the dimension of the feedforward network model (default=2048).
+        dropout: the dropout value (default=0.1).
+
+    Examples::
+        >>> transformer_model = nn.Transformer(src_vocab, tgt_vocab)
+        >>> transformer_model = nn.Transformer(src_vocab, tgt_vocab, nhead=16, num_encoder_layers=12)
+    """
+
+    def __init__(self, d_model=512, nhead=8, num_encoder_layers=6,
+                 num_decoder_layers=6, dim_feedforward=2048, dropout=0.1):
+        encoder_layer = TransformerEncoderLayer(d_model, nhead, dim_feedforward, dropout)
+        encoder_norm = LayerNorm(d_model) 
+        encoder = TransformerEncoder(encoder_layer, num_encoder_layers, encoder_norm)
```

Comment:
We should make encoder and decoder keyword arguments and create this if they're set to None. This allows users to pass in custom code.